core/common/src/main/java/alluxio/resource/DynamicResourcePool.java

```diff
@@ -0,0 +1,481 @@
+          closeResource(resource);
+        }
+      }
+    }, options.getInitialDelayMs(), options.getGcIntervalMs(), TimeUnit.MILLISECONDS);
+  }
+
+  /**
+   * Acquire a resource of type {code T} from the pool.
+   *
+   * @return the acquired resource
+   * @throws IOException if it fails to acquire because of the failure to create a new resource
+   */
+  @Override
+  public T acquire() throws IOException {
+    try {
+      return acquire(100  /* no timeout */, TimeUnit.DAYS);
+    } catch (TimeoutException e) {
+      LOG.error("Never should timeout in acquire().");
+      throw new RuntimeException(e);
+    }
+  }
+
+  /**
+   * Acquires a resource of type {code T} from the pool.
+   *
+   * This method is like {@link #acquire()}, but it will time out if an object cannot be
+   * acquired before the specified amount of time.
+   *
+   * @param time an amount of time to wait, null to wait indefinitely
+   * @param unit the unit to use for time, null to wait indefinitely
+   * @return a resource taken from the pool
+   * @throws IOException if it fails to acquire because of the failure to create a new resource
+   * @throws TimeoutException if it fails to acquire because of time out
+   */
+  @Override
+  public T acquire(long time, TimeUnit unit) throws IOException, TimeoutException {
+    long endTimeMs = System.currentTimeMillis() + unit.toMillis(time);
+
+    // Try to take a resource without blocking
+    ResourceInternal<T> resource = poll();
+    if (resource != null) {
+      if (isHealthy(resource.mResource)) {
+        return resource.mResource;
+      } else {
+        LOG.info("Clearing unhealthy resource {}.", resource.mResource.toString());
+        closeResource(resource.mResource);
+        remove(resource.mResource);
+        return acquire(time, unit);
+      }
+    }
+
+    if (!isFull()) {
+      // If the resource pool is empty but capacity is not yet full, create a new resource.
+      T newResource = createNewResource();
+      ResourceInternal<T> resourceInternal = new ResourceInternal<>(newResource);
+      // It is possible that we have more resources than mMaxCapacity here, just insert it
+      // instead of waiting.
+      add(resourceInternal);
+      return newResource;
+    }
+
+    // Otherwise, try to take a resource from the pool, blocking if none are available.
+    try {
+      mLock.lockInterruptibly();
+      try {
+        while (true) {
+          resource = poll();
+          if (resource != null) {
+            break;
+          }
+          long currTimeMs = System.currentTimeMillis();
+          if (currTimeMs >= endTimeMs || !mNotEmpty
+              .await(endTimeMs - currTimeMs, TimeUnit.MILLISECONDS)) {
+            throw new TimeoutException("Acquire resource times out.");
+          }
+        }
+      } finally {
+        mLock.unlock();
+      }
+    } catch (InterruptedException e) {
+      throw new RuntimeException(e);
```

Comment:
done.

I don't see a strong reason to use lock interruptibly. I copied what the ResourcePool did. Removed.